### PR TITLE
Implement ObservableValue<T>.asFlow()

### PR DESCRIPTION
**Pull request by dkhalanskyjb**
Solves https://github.com/Kotlin/kotlinx.coroutines/issues/1695

It seems there is no way to implement this for `Observable` as it doesn't even expose the type of the observable value.

### CODE REVIEW

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+@ExperimentalCoroutinesApi
+fun <T: Any> ObservableValue<T>.asFlow(): Flow<T> = callbackFlow<T> {
+    val listener = ChangeListener<T> { observable, oldValue, newValue ->
+        try {
+            offer(newValue)
+        } catch (e: CancellationException) {
+            // In case the event fires after the channel is closed
+        }
+    }
+    withContext(Dispatchers.JavaFx) {
```
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Also, could you please port [this test](https://github.com/ReactiveX/RxJavaFX/blob/a78ca7d15f7d82d201df8fafb6eba732ec17e327/src/test/java/io/reactivex/rxjavafx/RxJavaFXTest.java) to `ObservableValue<T>.asFlow()`? It looks like we are missing a decent integration test

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Since this implementation uses [ObservableValue.addListener], even if this [ObservableValue]
+ * supports lazy evaluation, eager computation will be enforced while the flow is being collected.
+ */
+@ExperimentalCoroutinesApi
+fun <T: Any> ObservableValue<T>.asFlow(): Flow<T> = callbackFlow<T> {
+    val listener = ChangeListener<T> { observable, oldValue, newValue ->
+        try {
+            offer(newValue)
+        } catch (e: CancellationException) {
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: What is going to happen if  `offer` throws a non-cancellation exception, e.g. because `collect {}` block has thrown it? Is it behaviour consistent and non-racy?

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: > Is it necessary to add and remove listener in JavaFx thread?

It turns out that JavaFX is supposed to be used in a single thread for *everything*. It did catch me, too.

> Could you please point out why?

Sure: `removeListener` and `set` are not synchronized at all, and, as seen here https://java-browser.yawk.at/org.openjfx/javafx-base/11/com/sun/javafx/binding/ExpressionHelper.java#348, and here https://java-browser.yawk.at/org.openjfx/javafx-base/11/com/sun/javafx/binding/ExpressionHelper.java#311, if `removeListener` is run somewhere other than JavaFX thread, an NPE or an OutOfRange could happen.

Likewise with `addListener`: if another `addListener` was running in the JavaFX thread, similar problems would occur.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            // In case the event fires after the channel is closed
+        }
+    }
+    withContext(Dispatchers.JavaFx) {
+        addListener(listener)
+        send(value)
```
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: No one would dare to send a value in a way that could race with an `addListener`, because it could lead to an NPE, for example. I could describe a specific execution if needed, but that is highly elaborate and relies on happens-before reasoning. My point is, JavaFX is really not thread-safe, so if someone tries to do something outside of the JavaFX thread, they are bound to find trouble no matter what.

File: ui/kotlinx-coroutines-javafx/test/JavaFxObservableAsFlowTest.kt [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+
+        flow.collect { i ->
+            assertTrue(i == 1 || i == 2)
+        }
+    }
+
+    @Test
+    fun cancellationRaceStressTest() = runTest {
```
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/test/JavaFxObservableAsFlowTest.kt]: What constitutes a stress-test? Many tests in this code base use `stressTestMultiplier` without being called stress-tests. Should I rename this test not to have `StressTest` in the name or should I move it to a separate file? What is the difference?

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * The resulting flow is conflated, meaning that if several values arrive in a quick succession, only
+ * the last one will be produced.
+ *
+ * It produces at least one value.
+ *
+ * Since this implementation uses [ObservableValue.addListener], even if this [ObservableValue]
+ * supports lazy evaluation, eager computation will be enforced while the flow is being collected.
+ */
+@ExperimentalCoroutinesApi
+fun <T: Any> ObservableValue<T>.asFlow(): Flow<T> = callbackFlow<T> {
```
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Understood, thanks. I told IDEA not to chastise me for redundant `public` qualifiers, and will fix this.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 2 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: As usual for this PR, I think that this is fine because everything here (this offer here, `addListener`, `removeListener` and so on) happens in the same JavaFX thread. So it's probably fully consistent and not racy.

I found a problem though: if someone runs `collect` in the JavaFX main thread, everything hangs. Should we consider this case? And can we?

Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Removing `runBlocking`, as recommended above, did solve the hanging problem.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Fixed.

File: ui/kotlinx-coroutines-javafx/test/JavaFxObservableAsFlowTest.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/test/JavaFxObservableAsFlowTest.kt]: Fixed.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: This version does work nicely! I replaced mine with it.

I have not yet understood how to adapt the test though. I am submitting a new version of code for review anyway because introducing a new test is independent from other changes.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+public fun <T: Any> ObservableValue<T>.asFlow(): Flow<T> = callbackFlow<T> {
+    val listener = ChangeListener<T> { _, _, newValue ->
+        try {
+            offer(newValue)
+        } catch (e: CancellationException) {
+            // In case the event fires after the channel is closed
+        }
+    }
+    addListener(listener)
+    send(value)
```
Comment (dkhalanskyjb) [inline ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: There is a problem with this behavior:
```kotlin
    @Test
    fun failure() = runTest {
        withContext(Dispatchers.JavaFx) {
            val END_MARKER = "xxx"
            val integerProperty = SimpleStringProperty()
            val flow = integerProperty.asFlow().takeWhile { j -> j != END_MARKER }
            launch {
                yield()
                yield()
                integerProperty.set("xxx")
            }
            flow.collect {

            }
        }
    }
```

Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: This test fails, because the initial value is `null`, and `null` can't be passed to `send` for some reason.

However, the problem is deeper:
```kotlin
    @Test
    fun failure() = runTest {
        withContext(Dispatchers.JavaFx) {
            val END_MARKER = "xxx"
            val integerProperty = SimpleStringProperty("xyzzy")
            val flow: Flow<String?> = integerProperty.asFlow().takeWhile {
                it != null && it == END_MARKER
            }
            launch {
                yield()
                integerProperty.set(null)
                yield()
                integerProperty.set(END_MARKER)
            }
            flow.collect {

            }
        }
    }
```
also fails, and this can't be easily fixed: https://docs.oracle.com/javase/8/javafx/api/javafx/beans/value/WritableIntegerValue.html#setValue-java.lang.Number- JavaFX is, as seen in many places, fine with storing `null` in properties.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Added the integration test.

File: ui/kotlinx-coroutines-javafx/test/examples/FxAsFlow.kt [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+ * https://github.com/ReactiveX/RxJavaFX/blob/a78ca7d15f7d82d201df8fafb6eba732ec17e327/src/test/java/io/reactivex/rxjavafx/RxJavaFXTest.java
+ */
+class FxAsFlowApp: Application(), CoroutineScope {
+
+    private var job = Job()
+    override val coroutineContext: CoroutineContext
+        get() = JavaFx + job
+
+    private  val  incrementBttn = Button("Increment")
+    private  val  incrementLabel = Label("")
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/test/examples/FxAsFlow.kt]: Please format it properly

File: ui/kotlinx-coroutines-javafx/test/JavaFxStressTest.kt [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+class JavaFxStressTest : TestBase() {
+
+    @Before
+    fun setup() {
+        ignoreLostThreads("JavaFX Application Thread", "Thread-", "QuantumRenderer-", "InvokeLaterDispatcher")
+    }
+
+    @Test
+    fun cancellationRaceStressTest() = runTest {
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/test/JavaFxStressTest.kt]: `testCancellationRace`: `test` should be in the prefix and `stressTest` is already present in the class name

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 6 above)
Comment (qwwdfsad) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: This test does not pass because the compiler inserts null-checks in `ChangeListener`.
And it does so because you have explicit `Any` bound: `public fun <T: Any> ObservableValue<T>.asFlow()`.

It looks like you can remove `: Any` upper bound and ignore the initial null value.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
+package kotlinx.coroutines.javafx
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Missing copyright. You can setup one in `Preferences -> Copyright profiles`:
`Copyright 2016-$today.year JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.`

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.conflate
+import kotlinx.coroutines.flow.flowOn
+
+/**
+ * Creates an instance of a cold [Flow] that subscribes to the given [ObservableValue] and produces
+ * its values as they change.
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: It's worth to mention that is all happening on Fx main thread.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * The resulting flow is conflated, meaning that if several values arrive in a quick succession, only
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: It's worth to mention that it is possible to change this behaviour with `[buffer]` operator, similarly to other operators

File: ui/kotlinx-coroutines-javafx/test/JavaFxObservableAsFlowTest.kt [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Test
+    fun testFlowOrder() = runTest {
+        if (!initPlatform()) {
+            println("Skipping JavaFxTest in headless environment")
+            return@runTest // ignore test in headless environments
+        }
+
+        val integerProperty = SimpleIntegerProperty(0)
+        val n = 10000 * stressTestMultiplier
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/test/JavaFxObservableAsFlowTest.kt]: please remove stressTestMultiplier and make the constant smaller

**Comment by qwwdfsad** [comment]
Please, rebase all the changes on current develop

File: ui/kotlinx-coroutines-javafx/test/examples/FxAsFlow.kt (hunk 7 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/test/examples/FxAsFlow.kt]: Done.

File: ui/kotlinx-coroutines-javafx/test/JavaFxStressTest.kt (hunk 8 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/test/JavaFxStressTest.kt]: Fixed.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 6 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Yes, that helped!

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 9 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Configured and fixed.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 10 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: I added a mention, but depending on why it's needed, maybe it should be worded differently.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 11 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Added some text about operator fusion.

File: ui/kotlinx-coroutines-javafx/test/JavaFxObservableAsFlowTest.kt (hunk 12 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/test/JavaFxObservableAsFlowTest.kt]: Done.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+package kotlinx.coroutines.javafx
+
+import javafx.beans.value.ChangeListener
+import javafx.beans.value.ObservableValue
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.*
+
+/**
+ * Creates an instance of a cold [Flow] that subscribes to the given [ObservableValue] and produces
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: KDoc wording: Flow does not *produce*, it *emits*

Comment (qwwdfsad) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: `a quick succession` -> `quick succession`

Comment (qwwdfsad) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: A lot of newlines are not really necessary

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 10 above)
Comment (qwwdfsad) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: Please use `Dispatchers.JavaFx` wording instead of `in the JavaFX application thread.`
Referring library primitives is always beneficial

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+ * its values as they change.
+ *
+ * The resulting flow is conflated, meaning that if several values arrive in a quick succession, only
+ * the last one will be produced.
+ *
+ * It produces at least one value.
```
Comment (qwwdfsad) [inline ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: >It produces at least one value.

This guarantee is a bit weird and leads to an inconvenient API.
Basically, it means that one **cannot** have `ObservableValue<ObjectSubtype>` without a default value (a common case in UI) and non-null bound.
Have you considered an option to ignore the very first value (line 42) if it is a `null` like Rx adapter does?

Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: >  Basically, it means that one cannot have `ObservableValue<ObjectSubtype>` without a default value (a common case in UI) and non-null bound.

True, but I don't see how this would be possible in any case:
https://docs.oracle.com/javase/8/javafx/api/javafx/beans/value/ObservableValue.html#getValue-- is it even allowed to implement an `ObservableValue` that throws an exception on calls to `getValue()`?

> This guarantee is a bit weird and leads to an inconvenient API.

The reason why I even added this guarantee was convenience: this would allow one to collect this flow in a thread aside from the JavaFX main thread and always have an initial value to work with, not worrying with synchronization. Otherwise, in order to always have an up-to-date version of an `ObservableValue` in a separate thread, one would have to block the JavaFX thread, which is a no-go.

>  Have you considered an option to ignore the very first value (line 42) if it is a null like Rx adapter does?

This would violate the guarantee that at least one value is produced, and if that guarantee does not hold, what would be a good reason to emit a value at all?

In short, if there is such a use case—collecting from a separate thread and being sure that you are not terribly out-of-date—then, I believe, at least one value should be emitted; otherwise, I don't see why emit an initial value in any case, but the solution with sending an initial value only if it is not null is seemingly not worse.

File: ui/kotlinx-coroutines-javafx/test/JavaFxStressTest.kt [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        if (!initPlatform()) {
+            println("Skipping JavaFxTest in headless environment")
+            return@runTest // ignore test in headless environments
+        }
+
+        val integerProperty = SimpleIntegerProperty(0)
+        val flow = integerProperty.asFlow()
+        var i = 1
+        val n = 1000 * stressTestMultiplier
+        newSingleThreadContext("collector").use { pool ->
```
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/test/JavaFxStressTest.kt]: Could you please elaborate? grepping the whole codebase for `TestExecutor` only shows something seemingly irrelevant.

File: ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt (hunk 13 above)
Comment (dkhalanskyjb) [reply ui/kotlinx-coroutines-javafx/src/JavaFxConvert.kt]: I took the word "produce" from https://kotlinlang.org/docs/reference/coroutines/flow.html where it is used not only in the context of producing an output of the whole program, so maybe it, too, should be fixed there.

In any case, I committed an improved KDoc.

File: ui/kotlinx-coroutines-javafx/test/JavaFxStressTest.kt (hunk 15 above)
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply ui/kotlinx-coroutines-javafx/test/JavaFxStressTest.kt]: My bad, meant `ExecutorRule`
